File: Medium/Java/LinkedList/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode res = new ListNode(0);
        ListNode p = res;
        while (l1 != null || l2 != null) {
            int l1val = 0, l2val = 0;
            if (l1 != null)
                l1val = l1.val;
            if (l2 != null)
                l2val = l2.val;
            int temp = l1val + l2val + carry;
            if (temp >= 10) {
                temp -= 10;
                carry = 1;
            }
            else
                carry = 0;
            p.next = new ListNode(temp);
            p = p.next;
            if (l1 != null)
                l1 = l1.next;
            if (l2 != null)
                l2 = l2.next;
        }
        if (carry == 1)
            p.next = new ListNode(1);
        return res.next;
    }
}

File: Medium/Python/LinkedList/AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
            :type l1: ListNode
            :type l2: ListNode
            :rtype: ListNode
            """
        # elegant version
        carry = 0
        root = n = ListNode(0)
        while l1 or l2 or carry:
            v1 = v2 = 0
            if l1:
                v1 = l1.val
                l1 = l1.next
            if l2:
                v2 = l2.val
                l2 = l2.next
            carry, val = divmod(v1 + v2 + carry, 10)
            n.next = ListNode(val)
            n = n.next
        return root.next

        # ugly
#         if not l1 or not l2:
#             return l1 if l1 else l2

#         carry = False
#         result = ListNode(l1.val + l2.val)
#         if result.val >= 10:
#             result.val -= 10
#             carry = True
#         p = result; p1 = l1.next; p2 = l2.next

#         while p1 and p2:
#             p.next = ListNode(p1.val + p2.val)
#             p = p.next
#             p1 = p1.next
#             p2 = p2.next
#             if carry:
#                 p.val += 1
#             if p.val >= 10:
#                 p.val -= 10
#                 carry = True
#             else:
#                 carry = False

#         if p1:
#             p.next = p1
#         if p2:
#             p.next = p2
#         while carry:
#             if p.next == None:
#                 p.next = ListNode(1)
#                 carry = False
#             else:
#                 p = p.next
#                 p.val += 1
#                 if p.val >= 10:
#                     p.val -= 10
#                     carry = True
#                 else:
#                     carry = False

#         return result
